### PR TITLE
zapi: Add simple space crud commands

### PR DESCRIPTION
**Pull request by mattnibs**
Focusing on simple space crud operations, add these commands to
zapi:
- ls: list all spaces with option to glob search on space names
- new: create a space
- rm: delete a space
- rename: rename a space

building on #526

### CODE REVIEW

File: cmd/zapi/cmd/info/ls.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+		names := api.SpaceInfos(matches).Names()
+		width := termwidth.Width()
+		// print listing laid out in columns like ls
+		err := colw.Write(os.Stdout, names, width, 3)
+		if err == colw.ErrDoesNotFit {
+			fmt.Println(strings.Join(names, "\n"))
+		} else if err != nil {
+			return err
+		}
+	}
```
Comment (alfred-landrum) [inline cmd/zapi/cmd/info/ls.go:66]:
```suggestion
      }
      names := api.SpaceInfos(matches).Names()
      ...
```

File: cmd/zapi/cmd/rename/rename.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	*cmd.Command
+}
+
+func (c *Command) Run(args []string) error {
+	if len(args) > 2 {
+		return errors.New("too many arguments")
+	}
+	if len(args) < 2 {
+		return errors.New("not enough arguments")
+	}
```
Comment (alfred-landrum) [inline cmd/zapi/cmd/rename/rename.go:39]:
```suggestion
	if len(args) != 2 {
		return errors.New("expected <old_name> <new_name>")
	}
```

File: cmd/zapi/cmd/rename/rename.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	c.Spacename = args[0]
+	newname := args[1]
+	id, err := c.SpaceID()
```
Comment (alfred-landrum) [inline cmd/zapi/cmd/rename/rename.go:42]: It's seems odd to reuse the command logic to handle the spaceid option for this. How about breaking out the spaceid -> name translation to a separate function you can use here?

File: cmd/zapi/cmd/rename/rename.go (hunk 2 above)
Comment (mattnibs) [reply cmd/zapi/cmd/rename/rename.go:39]: much better!

File: cmd/zapi/cmd/rm/rm.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	}
+	if len(args) == 1 {
+		c.Spacename = args[0]
+	}
+	id, err := c.SpaceID()
+	if err != nil {
+		return err
+	}
+	err = c.Client().SpaceDelete(c.Context(), id)
+	if err != nil {
```
Comment (nwt) [inline cmd/zapi/cmd/rm/rm.go:42]:
```suggestion
	if err := c.Client().SpaceDelete(c.Context(), id); err != nil {
```

File: cmd/zapi/cmd/info/ls.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+// Run lists all spaces in the current zqd host or if a parameter
+// is provided (in glob style) lists the info about that space.
+func (c *LsCommand) Run(args []string) error {
+	client := c.Client()
+	matches, err := cmd.SpaceGlob(c.Context(), client, args...)
+	if err != nil {
+		if err == cmd.ErrNoSpacesExist {
+			fmt.Fprintln(os.Stdout, "no spaces exist")
```
Comment (nwt) [inline cmd/zapi/cmd/info/ls.go:45]:
```suggestion
			fmt.Println("no spaces exist")
```

File: cmd/zapi/cmd/new/new.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+	if len(args) > 1 {
+		return errors.New("too many arguments")
+	}
+	if len(args) != 1 {
+		return errors.New("must specify a space name")
+	}
+	client := c.Client()
+	name := args[0]
+	_, err := client.SpacePost(c.Context(), api.SpacePostRequest{Name: name})
+	if err != nil {
```
Comment (nwt) [inline cmd/zapi/cmd/new/new.go:42]:
```suggestion
	if _, err := client.SpacePost(c.Context(), api.SpacePostRequest{Name: name}); err != nil {
```

File: cmd/zapi/cmd/rename/rename.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+		return errors.New("expected <old_name> <new_name>")
+	}
+	oldname := args[0]
+	newname := args[1]
+	id, err := cmd.GetSpaceID(c.Context(), c.Client(), oldname)
+	if err != nil {
+		return err
+	}
+	err = c.Client().SpacePut(c.Context(), id, api.SpacePutRequest{Name: newname})
+	if err != nil {
```
Comment (nwt) [inline cmd/zapi/cmd/rename/rename.go:44]:
```suggestion
	if err := c.Client().SpacePut(c.Context(), id, api.SpacePutRequest{Name: newname}); err != nil {
```

File: cmd/zapi/cmd/info/ls.go (hunk 5 above)
Comment (mattnibs) [reply cmd/zapi/cmd/info/ls.go:45]: I think in an official cli like this it's good to be specific about where things are headed.

File: pkg/colw/colw.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+func Write(w io.Writer, in []string, width, pad int) error {
+	widths := Layout(in, width, pad)
+	if widths == nil {
+		return ErrDoesNotFit
+	}
+	ncol := len(widths)
+	padding := makePadding(width, ' ')
+	for k, s := range in {
+		_, err := w.Write([]byte(s))
+		if err != nil {
```
Comment (nwt) [inline pkg/colw/colw.go:80]:
```suggestion
		if _, err := w.Write([]byte(s)); err != nil {
```

File: pkg/colw/colw.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+			return err
+		}
+		if k == len(in)-1 {
+			_, err := w.Write(newline)
+			return err
+		}
+		colno := k % ncol
+		if colno == ncol-1 {
+			_, err := w.Write(newline)
+			if err != nil {
```
Comment (nwt) [inline pkg/colw/colw.go:90]:
```suggestion
			if _, err := w.Write(newline); err != nil {
```

File: pkg/colw/colw.go [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+				return err
+			}
+		} else {
+			width := widths[colno]
+			pad := width - len(s)
```
Comment (nwt) [inline pkg/colw/colw.go:95]:
```suggestion
			pad := widths[colno] - len(s)
```

File: pkg/colw/colw.go [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+			if pad > 0 {
+				_, err := w.Write(padding[0:pad])
+				if err != nil {
```
Comment (nwt) [inline pkg/colw/colw.go:98]:
```suggestion
				if _, err := w.Write(padding[0:pad]); err != nil {
```

File: cmd/zapi/cmd/info/ls.go (hunk 5 above)
Comment (alfred-landrum) [reply cmd/zapi/cmd/info/ls.go:45]: fmt.Println is very specific:
```go
func Println(a ...interface{}) (n int, err error) {
	return Fprintln(os.Stdout, a...)
}
```
Please change to use fmt.Println.

File: cmd/zapi/cmd/space.go [hunk 12]
```diff
@@ -40,6 +44,18 @@ func SpaceGlob(ctx context.Context, client *api.Connection, patterns ...string)
 	return spaces, nil
 }
 
+func GetSpaceID(ctx context.Context, client *api.Connection, name string) (api.SpaceID, error) {
```
Comment (alfred-landrum) [inline cmd/zapi/cmd/space.go:47]: 👍

File: cmd/zapi/cmd/rename/rename.go [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+	"github.com/mccanne/charm"
+)
+
+var Rename = &charm.Spec{
+	Name:  "rename",
+	Usage: "rename [old_name] [new_name]",
+	Short: "renames a space",
+	Long: `The rename command takes two arguments. The first arugment is the
+	current name of the space and the second arugment is the new name of the space.
+	`,
```
Comment (alfred-landrum) [inline cmd/zapi/cmd/rename/rename.go:19]:
```suggestion
	Long: `Renames a space, given the current space name and a desired new name.`,
```

File: cmd/zapi/cmd/info/ls.go (hunk 5 above)
Comment (mattnibs) [reply cmd/zapi/cmd/info/ls.go:45]: ok

**Comment by alfred-landrum** [comment]
@mattnibs : this looks ready except for the `go fmt` failure showing in CI, will you take a look?

**Comment by mattnibs** [comment]
> @mattnibs : this looks ready except for the `go fmt` failure showing in CI, will you take a look?

Done

**Pull request merged**